compiler/rustc_passes/src/stability.rs

```diff
@@ -777,35 +777,43 @@ impl Visitor<'tcx> for Checker<'tcx> {
             hir::ItemKind::Union(..) if !self.tcx.features().untagged_unions => {
                 let def_id = self.tcx.hir().local_def_id(item.hir_id);
                 let ty = self.tcx.type_of(def_id);
-                let (adt_def, substs) = match ty.kind() {
-                    ty::Adt(adt_def, substs) => (adt_def, substs),
-                    _ => bug!(),
-                };
-
-                // Non-`Copy` fields are unstable, except for `ManuallyDrop`.
-                let param_env = self.tcx.param_env(def_id);
-                for field in &adt_def.non_enum_variant().fields {
-                    let field_ty = field.ty(self.tcx, substs);
-                    if !field_ty.ty_adt_def().map_or(false, |adt_def| adt_def.is_manually_drop())
-                        && !field_ty.is_copy_modulo_regions(self.tcx.at(DUMMY_SP), param_env)
-                    {
-                        if field_ty.needs_drop(self.tcx, param_env) {
-                            // Avoid duplicate error: This will error later anyway because fields
-                            // that need drop are not allowed.
-                            self.tcx.sess.delay_span_bug(
-                                item.span,
-                                "union should have been rejected due to potentially dropping field",
-                            );
-                        } else {
-                            feature_err(
-                                &self.tcx.sess.parse_sess,
-                                sym::untagged_unions,
-                                self.tcx.def_span(field.did),
-                                "unions with non-`Copy` fields other than `ManuallyDrop<T>` are unstable",
-                            )
-                            .emit();
+                match ty.kind() {
+                    ty::Adt(adt_def, substs) => {
+                        // Non-`Copy` fields are unstable, except for `ManuallyDrop`.
+                        let param_env = self.tcx.param_env(def_id);
+                        for field in &adt_def.non_enum_variant().fields {
+                            let field_ty = field.ty(self.tcx, substs);
+                            if !field_ty
+                                .ty_adt_def()
+                                .map_or(false, |adt_def| adt_def.is_manually_drop())
+                                && !field_ty
+                                    .is_copy_modulo_regions(self.tcx.at(DUMMY_SP), param_env)
```

Comment:
I think this is the actual problem here. imo `is_copy_modulo_regions` doesn't make sense outside of codegen/mir building.

This is currently a stability hole as the following compiles:
```rust
#[derive(Clone)]
struct Bar<'a>(&'a ());

impl Copy for Bar<'static> {}

union Foo<'a> {
    field: u32,
    other_field: Bar<'a>,
}
```
For some reason we also explicitly require types to be copy when initializing a union, so the following does error:
```rust
fn test<'a>(q: &'a ()) {
    let bar = Bar(&q);
    let foo: Foo<'a> = Foo {
        other_field: bar,
    };
}
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=45ea9990bc47e6066abeec085a372b85

---

Does using the following also fix this ICE?
```rust
let trait_ref = TraitRef::new(copy_def_id, tcx.intern_substs(&[field_ty.into()]));

let is_copy = tcx.infer_ctxt().enter(|infcx| {
    infcx.predicate_must_hold_considering_regions(Obligation::new(
        ObligationCause::dummy(),
        param_env,
        trait_ref.without_const().to_predicate(tcx),
    ))
});
```

Reply:
I don't think this is the root issue, but I agree that this seems like a potential soundness hole. I would maybe just file an issue for followup.

Reply:
> I don't think this is the root issue

My understanding is that the root issue is calling `codegen_fulfill_obligations` outside of codegen. I am a bit confused about why I said that `is_copy_modulo_regions` is the issue here when there's a call to `normailize_erasing_regions` right below it - as you've said in https://github.com/rust-lang/rust/pull/82039#discussion_r695351139. I personally prefer to "correctly" deal with regions in `needs_drop` and similar methods, even if that is a more complex change.